File: ipc-parser/src/main/java/com/satori/deser/domain/DisconnectResponse.java
package com.satori.deser.domain;

import com.fasterxml.jackson.databind.JsonNode;
import com.satori.deser.DomainSpecificResponse;
import com.satori.deser.DomainSpecificResponseType;
import com.satori.deser.Ticket;
import com.satori.deser.VirtualConnectionType;

import java.util.Iterator;
import java.util.Optional;

public class DisconnectResponse implements DomainSpecificResponse {
    public VirtualConnectionType virtualConnectionType;
    public long implicated_cid;
    private Long target_cid;
    private Ticket ticket;

    @Override
    public DomainSpecificResponseType getType() {
        return null;
    }

    @Override
    public Optional<Ticket> getTicket() {
        return this.ticket != null ? Optional.of(this.ticket) : Optional.empty();
    }

    @Override
    public Optional<String> getMessage() {
        return Optional.empty();
    }

    @Override
    public Optional<DomainSpecificResponse> deserializeFrom(JsonNode infoNode) {
        // debug purposes. Iterator<JsonNode> values1 = infoNode.elements();
        JsonNode hLANp2s = infoNode.get("HyperLANPeerToHyperLANServer");
        if (hLANp2s != null) {
            // first value is the ticket, the second, the implicated CID
            Iterator<JsonNode> values = hLANp2s.elements();
            if (!values.hasNext()) {
                return Optional.empty();
            }

            Optional<Ticket> ticket_opt = Ticket.tryFrom(values.next().longValue());

            if (!values.hasNext()) {
                return Optional.empty();
            }

            this.implicated_cid = values.next().longValue();
            this.ticket = ticket_opt.orElse(null);
            this.virtualConnectionType = VirtualConnectionType.HyperLANPeerToHyperLANServer;
            return Optional.of(this);
        }

        JsonNode hLANp2p = infoNode.get("HyperLANPeerToHyperLANPeer");
        if (hLANp2p != null) {
            // first value is the ticket, the second, the implicated CID
            Iterator<JsonNode> values = hLANp2p.elements();
            if (!values.hasNext()) {
                return Optional.empty();
            }

            Optional<Ticket> ticket_opt = Ticket.tryFrom(values.next().longValue());

            if (!values.hasNext()) {
                return Optional.empty();
            }

            long implicated_cid = values.next().longValue();

            if (!values.hasNext()) {
                return Optional.empty();
            }

            this.implicated_cid = implicated_cid;
            this.target_cid = values.next().longValue();
            this.ticket = ticket_opt.orElse(null);
            this.virtualConnectionType = VirtualConnectionType.HyperLANPeerToHyperLANPeer;
            return Optional.of(this);
        }


        return Optional.empty();
    }

    // This value won't exist if the connection type is a p2s type
    public Optional<Long> getTargetCID() {
        return Optional.ofNullable(this.target_cid);
    }
}

File: flutterust/packages/satori_ffi_parser/lib/types/dsr/disconnect_response.dart
import 'package:optional/optional.dart';
import 'package:satori_ffi_parser/types/domain_specific_response_type.dart';
import 'package:satori_ffi_parser/types/ticket.dart';
import 'package:satori_ffi_parser/types/virtual_connection_type.dart';

import '../domain_specific_response.dart';
import '../u64.dart';

class DisconnectResponse extends DomainSpecificResponse {
  final u64 implicated_cid;
  final u64 icid;
  final u64 peer_cid;
  final Optional<Ticket> ticket;
  final VirtualConnectionType virtualConnectionType;

  DisconnectResponse._(this.virtualConnectionType, this.implicated_cid, this.icid, this.peer_cid, this.ticket);

  @override
  Optional<String> getMessage() {
    return Optional.empty();
  }

  @override
  Optional<Ticket> getTicket() {
    return this.ticket;
  }

  @override
  DomainSpecificResponseType getType() {
    return DomainSpecificResponseType.Disconnect;
  }

  /*
    HyperLANPeerToHyperLANServer(#[serde(serialize_with = "string")] u64, #[serde(serialize_with = "string")] u64),
    HyperLANPeerToHyperLANPeer(#[serde(serialize_with = "string")] u64, #[serde(serialize_with = "string")] u64, #[serde(serialize_with = "string")] u64)
  */
  static Optional<DomainSpecificResponse> tryFrom(
      Map<String, dynamic> infoNode) {
    return findFirstInKeys(infoNode).flatMap((vConnType) {
      print("VConn type: " + vConnType.toString());
      List<dynamic> leaf = infoNode[vConnType.toString().split(".").last];
      if (leaf.length < 2) {
        return Optional.empty();
      }

      Optional<Ticket> ticket = Ticket.tryFrom(leaf[0]);
      return u64.tryFrom(leaf[1]).flatMap((implicated_cid) {
        Optional<u64> peer_cid;
        u64 icid = u64.zero;
        if (vConnType == VirtualConnectionType.HyperLANPeerToHyperLANPeer) {
          if (leaf.length != 3) {
            return Optional.empty();
          }

          peer_cid = u64.tryFrom(leaf[2]);
        } else {
          peer_cid = Optional.of(u64.zero);
        }

        return peer_cid.flatMap((peer_cid) {
          return Optional.of(DisconnectResponse._(
              vConnType, implicated_cid, icid, peer_cid, ticket));
        });
      });
    });
  }
}
